src/listener.rs

```diff
@@ -0,0 +1,410 @@
+//! A listener in the queue of listeners.
+
+use crate::sync::atomic::{AtomicUsize, Ordering};
+use crate::sync::cell::Cell;
+
+use core::task::Waker;
+
+#[cfg(feature = "std")]
+use parking::Unparker;
+
+/// A listener in the queue of listeners.
+pub(crate) struct Listener {
+    /// The current state of this listener.
+    ///
+    /// This also serves as the refcount:
+    /// - Ref 1: The listener state is not orphaned.
+    /// - Ref 2: The listener is in the queue (queue bit is set).
+    state: AtomicUsize,
+
+    /// The waker or thread handle that will be notified when the listener is
+    /// ready.
+    ///
+    /// This is kept synchronized by the `state` variable; therefore, it's
+    /// technically `Sync`.
+    waker: Cell<Wakeup>,
```

Comment:
wouldn't it make more sense to use `UnsafeCell` here?

Reply:
I originally used `UnsafeCell`, but given that most of the semantics involved are just "replace the `waker` with something else", I ended up using `Cell` instead.

I tried replacing it with `UnsafeCell<MaybeUninit<Wakeup>>` (and also getting rid of the `Wakeup::None` variant), but any gains are minimal at best. I figure it's better to just use `Cell` and avoid the unsafe code.